src/main/java/org/jetbrains/bsp/bazel/server/ScalaBspServer.java

```diff
@@ -21,84 +21,132 @@
+                actionGraphParser.getInputsAsUri(target, execRoot),
+                target))
+        .collect(Collectors.toList());
+  }
+
+  private Stream<ScalacOptionsItem> collectScalacOptionsResult(ActionGraphParser actionGraphParser,
+      List<String> options, List<String> inputs, String target) {
+    List<String> suffixes = ImmutableList.of(".jar", ".js");
+
+    return actionGraphParser.getOutputs(target, suffixes).stream()
+        .map(output -> buildScalacOptionsItem(options, inputs, target, output));
+  }
+
+  private ScalacOptionsItem buildScalacOptionsItem(List<String> options, List<String> inputs, String target,
+      String output) {
+    BuildTargetIdentifier buildTargetIdentifier = new BuildTargetIdentifier(target);
+    String execRootUri = Uri.fromExecPath("exec-root://" + output, execRoot).toString();
+
+    return new ScalacOptionsItem(buildTargetIdentifier, options, inputs, execRootUri);
   }
 
-  public CompletableFuture<ScalaTestClassesResult> buildTargetScalaTestClasses(
+  public Either<ResponseError, ScalaTestClassesResult> buildTargetScalaTestClasses(
       ScalaTestClassesParams scalaTestClassesParams) {
-    System.out.printf("DWH: Got buildTargetScalaTestClasses: %s%n", scalaTestClassesParams);
-    // TODO(illicitonion): Populate
-    return CompletableFuture.completedFuture(new ScalaTestClassesResult(new ArrayList<>()));
+    Build.QueryResult query = bazelQueryRunner.queryAllTargets();
+    ScalaTestClassesResult scalaTestClassesResult =
+        new ScalaTestClassesResult(
+            query.getTargetList().stream()
+                .map(Build.Target::getRule)
+                .filter(
+                    rule ->
+                        scalaTestClassesParams.getTargets().stream()
+                            .anyMatch(target -> target.getUri().equals(rule.getName())))
+                .filter(
+                    rule ->
+                        rule.getAttributeList().stream()
+                            .anyMatch(
+                                attribute ->
+                                    attribute.getName().equals("main_class")
+                                        && attribute.hasExplicitlySpecified()
+                                        && attribute.getExplicitlySpecified()))
+                .map(
+                    rule ->
+                        new ScalaTestClassesItem(
+                            new BuildTargetIdentifier(rule.getName()),
+                            getTestClasses(rule, query)))
+                .collect(Collectors.toList()));
+    return Either.forRight(scalaTestClassesResult);
+  }
+
+  private List<String> getTestClasses(Build.Rule rule, Build.QueryResult queryResult) {
+    return rule.getAttributeList().stream()
+        .filter(
+            attribute ->
+                attribute.getName().equals("main_class")
+                    && attribute.hasExplicitlySpecified()
+                    && attribute.getExplicitlySpecified())
```

Comment:
This condition is also used above. Perhaps it can be extracted?

Reply:
of course